### DISCUSSION TITLE
Clipping to discrete shape

### DISCUSSION THREAD

**Discussion by Rycha2021** [iTwin/itwinjs-core#2585] [Q&A] [answered]
Is it possible to clip to a discrete 3D shape like, for example, a Pipe meandering across an asset?
Currently in the demo, someone can move a cube clipping boundaries but when there is a 3D element like this, the cube still includes everything else. 
![Clip01](https://user-images.githubusercontent.com/86642020/138805958-4185be73-30ae-45cf-91a9-b45ce34941fe.jpg)

Is it possible **just** to select the pipe (i think you call it maybe a classifier?) and have tight clipping around an object like this complete piping run and not include anything else except the pipe? Else makes explaining this, other than highlighting something, less useful.

Hope this is possbile.

Ry

**Comment by pmconne** [comment]
Please clarify the result you are trying to achieve.
Do you want to:
- Only see the one pipe in the view? Or
- Project the shape of the pipe onto the XY plane and only draw geometry that intersects the volume extruded vertically from that footprint?

**Comment by Rycha2021** [comment]
Thank-you for responding.

Yes, in this example, I only wish to see the complete length of pipe isolated on its own and not a complete cube around the pipe at the X,Y and Z extremities with other elements therein.

From what I can see, isolating a single 3D element that could have complex X,Y and Z length will only be allowed if the entire contents of the bounding cube is shown.

![Picture2](https://user-images.githubusercontent.com/86642020/138862318-65877e82-8219-42d0-aa4f-9e803e6c78c5.png)

What if i have a system with multiple pipes in a pipe rack extending across a Plant. By the time we select the pipe, the entire plant may be visible in the cube formed. 

I presume the envelope, rather than a cube, would follow the centreline of the pipe and have a diameter from the centreline to enable only the pipe to be viewed?

In CAD model this is possible. I hope this is possible in iTwin. Thanks in advance.

Ry

**Comment by pmconne** [comment]
Isolating elements does not require clipping (and a clip volume produced from your pipe's geometry would be quite expensive to apply to the view).
Use [Viewport.setAlwaysDrawn](https://www.itwinjs.org/reference/imodeljs-frontend/views/viewport/setalwaysdrawn/) or [EmphasizeElements.isolateElements](https://www.itwinjs.org/reference/imodeljs-frontend/rendering/emphasizeelements/isolateelements/) or [EmphasizeElements.isolateSelectedElements](https://www.itwinjs.org/reference/imodeljs-frontend/rendering/emphasizeelements/isolateelements/).

**Comment by Rycha2021** [comment]
Unsure what you mean as "expensive". Please elaborate.

If I wanted to use this shape as a classifier wherein a reality model is clipped to follow the pipe and not have a cube, is this possible?

Currently, if i have a vessel in a plane, this clips the reality model OK as the cube can wrap the shape. When i have X,Y and Z (this example a pipe) over a broad area, i get the pipe and everything else in the cube. I only want the pipe shown regardless of shape and extent. Imagine the example image above as a reality model clip using a classifier or similar.......

Thanks.

Ry

**Reply by bbastings** [reply]
Are you essentially looking to create a "2d clip shape" from a 3d element's "shadow" on the ground? Something like the following:
![image](https://user-images.githubusercontent.com/65233531/138879908-e0238f4f-600b-4d98-a443-3405c3c472f9.png)
In the above image I isolated the "pipe" geometry and just drew a sloppy clip shape in a top view with the camera off (you can create non-rectangular clips by holding down the ctrl key to specify additional points when using the create clip shape tool).

**Reply by Rycha2021** [reply]
Thank-you for your response. I was not aware of this feature of CTRL. Cool. Not exactly what i am seeking however.

Ry

**Comment by pmconne** [comment]
> Unsure what you mean as "expensive". Please elaborate.

A clip is defined as a set of planes. The more complex the geometry from which you create the clip, the more planes you end up with. The more planes there are, the more work your GPU must perform to apply the clip, resulting in reduced framerate.

> If I wanted to use this shape as a classifier wherein a reality model is clipped to follow the pipe and not have a cube, is this possible?

Yes, that is why I asked you to clarify what you want, to which you replied "I only wish to see the complete length of pipe isolated on its own". Please be clearer.

**Comment by pmconne** [comment]
If you are trying to clip a **reality model** or map imagery based on the footprints of one or elements, you probably want to use a [planar clip mask](https://www.itwinjs.org/reference/imodeljs-common/displaystyles/planarclipmasksettings/).

**Reply by Rycha2021** [reply]
Thanks Paul. 

Guidance therein states “Currently reality models (including background maps and terrain) can be masked, but design models can only produce masks.”  

Is the last part a reference to a CAD element of similar shape acting as a classifier (mask)?

Any examples in a similar application or code?

Thanks 

Ry

**Reply by pmconne** [reply]
> Is the last part a reference to a CAD element of similar shape acting as a classifier (mask)?

Yes. You can use a CAD (design) model, or elements thereof, to produce a clip mask that can be used to clip reality models or maps.
You cannot currently apply the clip mask to a design model.

> Any examples in a similar application or code?

Maybe - can you please answer my question - is "use an element to mask out portions of a reality mesh" the thing you are trying to accomplish?

**Reply by Rycha2021** [reply]
Ideally yes in answer to your last question though not having much success with generating a good quality mesh model but that will come with more practice.

What i need though is the ability for the clip tool to follow a classifier (?correct term) **tightly**. In trial, if i select a classifier within the reality model, yes the clip masks a portion of the reality model but it is within a cube that also shows everything else inside. Not what i seek.

If you look at the pipe example in the previous post. If that was the classifier within a reality model, i want only the portions of the reality model to show exactly (within tolerance) to the classifier. All I manage to get currently is the first post where everything is shown in the clip cube.

I am not too sure how much more i can explain the solution sought.

It just appears that clipping to a tight classifier within a reality model is not possible or am i missing some functionality that could make this work?

Thanks again.

Ry

**Reply by pmconne** [reply]
Show your code, share your model.

**Comment by Rycha2021** [comment]
Part of model showing complexity.
![Picture5](https://user-images.githubusercontent.com/86642020/138980069-251ca164-7193-4886-9452-089530002de2.jpg)

Clipping to each pipe is sought.

Ry

**Comment by pmconne** [comment, marked as answer]
@Rycha2021 here is [a sandbox](https://www.itwinjs.org/sandboxes/pmconne/Mask%20reality%20model%20by%20element) demonstrating how to do the inverse: it clips out only those portions of the reality model that intersect the footprint of a given element.
It uses [PlanarClipMaskMode.IncludeElements](https://www.itwinjs.org/reference/imodeljs-common/displaystyles/planarclipmaskmode/) to produce the mask.
You would think that PlanarClipMaskMode.**Exclude**Elements would produce the effect you are looking for, but in general it won't, because it produces a mask from all elements except the one you specify - so, for example, the floor in your first screenshot is going to mask out the areas you want to see.
I'm looking into a way to produce the effect you want. I don't think it will be difficult.

**Reply by RBBentley** [reply]
I believe it may be possible to have the planar clip masks clip outside of a set of elements rather than inside.   I'll try to mock something up and see if it works and will report back next week.

**Reply by Rycha2021** [reply]
@pmconne and @RBBentley thank-you for your efforts and support here.

In the sandbox i can see the intent. 

I was hoping this would be less complicated than what it seems to be. Clipping, where in, the object appears within a cube works well. Tightly clipping to solely the element so that only the element appears in the reality model will offer good visualisation benefits in crowded and build-up assets / environments to isolate and rotate around specific elements.

Thank-you again and will look out for your learned guidance.

Ry.

**Reply by RBBentley** [reply]
I have implemented an additional PlanarClipMask property [here](https://github.com/iTwin/itwinjs-core/commit/f006473c0009f852321acfad976a7b59b9bda8fe)  to allow planar clip masks to invert their sense and clip outside rather than inside.   This should allow you to display the reality model only within the footprint of the masking geometry.  See the planar clip mask tools that have a new invert argument [here](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend-devtools/src/tools/PlanarMaskTools.ts)

**Reply by ovaisa** [reply]
This capability is also a recurring request from our clients and is high on our priority.  Are these  updates now available in release version, if not please let me know when they are expected to be available. 

Thanks,

Ovais

**Reply by pmconne** [reply]
The feature is available now in the latest 3.0 dev packages. We expect to produce a release candidate for 3.0 this week, although the US holiday may push it back to next week. Official 3.0 release is expected shortly thereafter. You are welcome to test it out with the dev or release candidate packages (assuming you are running your own backend which you'll upgrade to 3.0), or wait for official release.

For what it's worth: we usually release a new minor version of the packages every 3 weeks. That release schedule has been suspended while we work on the new major version (3.0) - 2.x has been receiving primarily only bug fix releases. The schedule will resume once 3.0 is officially released, resulting in much shorter delays before new features become available.

**Reply by ovaisa** [reply]
I would be very interested in test 3.0, can you please forward any details on how I can access it and associated cra template (as I would like to start by creating a vanilla 3.0 app).  If there is any info on migrating 2.x apps, please forward (I am on 2.19.20).

Thanks,

Ovais

**Reply by pmconne** [reply]
If you're using a cra template then it sounds like you would be using the general purpose backend instead of running your own backend, in which case you'll need to await official release when a 3.0 GPB will be deployed. @kckst8 or @aruniverse do you have info regarding 3.0 cra templates?
These [changelogs](https://github.com/iTwin/itwinjs-core/blob/master/docs/changehistory/NextVersion.md) for 3.0 will be cleaned up and better organized before official release but they offer a preview.
[This tool](https://github.com/iTwin/codemods) will automate portions of the migration.

**Reply by aruniverse** [reply]
We should have an update on the 3.0 template by next week and hopefully some migration guide. I'll update once available

**Reply by Rycha2021** [reply]
@aruniverse, is 3.0 template far away?

**Reply by aruniverse** [reply]
No we're pretty close, I should hopefully have an update for you later today.

**Reply by Rycha2021** [reply]
> The feature is available now in the latest 3.0 dev packages. We expect to produce a release candidate for 3.0 this week, although the US holiday may push it back to next week. Official 3.0 release is expected shortly thereafter. You are welcome to test it out with the dev or release candidate packages (assuming you are running your own backend which you'll upgrade to 3.0), or wait for official release.
> 
> For what it's worth: we usually release a new minor version of the packages every 3 weeks. That release schedule has been suspended while we work on the new major version (3.0) - 2.x has been receiving primarily only bug fix releases. The schedule will resume once 3.0 is officially released, resulting in much shorter delays before new features become available.

@aruniverse and @pmconne , please absolve the belated response as i have had some other issues to resolve and Xmas holidays etc but considering 3.0 is now released, can i confirm this feature of clipping discrete shapes (e.g. like a a complex pipe in multi directions) is now possible? Examples of how to implement is greatly appreciated.

Thanks

Ry

**Reply by pmconne** [reply]
@Rycha2021 yes, the feature is included in 3.0.0. I have upgraded the sandbox previously provided to use iTwin.js 3.0.0 and turned on the `invert` flag for the planar clip mask.
This should in theory demonstrate the new feature, but it appears the sandbox app is failing to supply a Bing Maps access key so the effect is currently not visible.

**Comment by Rycha2021** [comment]
@aruniverse, is 3.0 template far away? 2 weeks has passed since your last update.

**Reply by aruniverse** [reply]
Hi @Rycha2021 , I'm really sorry for the delay in getting back to you. We're very close! Just trying to iron out a few last things. We're hoping to have a public beta release candidate of iTwin.js 3.0 and the iTwin Viewer built on it available by this Friday, with an official release early next year. I will update you as soon as we have made the public beta available.

**Reply by aruniverse** [reply]
See https://github.com/iTwin/itwinjs-core/discussions/2942

**Comment by pmconne** [comment]
@Rycha2021 here is [a sandbox example](https://www.itwinjs.org/sandboxes/pmconne/Inverted%20model%20mask) based on 3.0 demonstrating the inverse mask feature.

**Comment by Rycha2021** [comment]
Thanks @pmconne. I checked the sandbox but have the following error... "Error: Failed to obtain WebGL context".

Ry

**Reply by pmconne** [reply]
Use [this page](https://connect-imodeljscompatibility.bentley.com) to verify your browser is configured for webgl.
If the problem is not obvious, post a screenshot of the contents of that page.

**Comment by Rycha2021** [comment]
@pmconne i can now see the model.
But i am unsure how this helps the use case i am looking for? I have tried to select individual walls but i am unable to get the effect of what i originally posted? (refer below again). I dont wish to see a cube but a close match to a specific shape.

Perhaps can you record a short video of what you consider you have developed so i can see your approach? Perhaps the issue is my misunderstanding of your approach.

![image](https://user-images.githubusercontent.com/86642020/152073914-ae24f5b3-73ca-48db-934c-ca1806d084b4.png)

Thank-you.

Ry

**Reply by pmconne** [reply]
What cube? Did you read the code and/or [the documentation](https://www.itwinjs.org/reference/core-common/displaystyles/planarclipmasksettings/) of the feature we have implemented?
The two elements hilited below are used to create the mask. Observe the background map is clipped to their precise shapes.
![image](https://user-images.githubusercontent.com/22944042/152091663-daf126f5-e7df-4855-a5ec-0b02692c4901.png)

**Comment by tonygem2022** [comment]
Hello. I am new to the forum so please bear with me.

I noted this query and this is also exactly what we are looking to do but i also note this query: https://github.com/iTwin/itwinjs-core/issues/3860

The functionality we are seeking is a 3D mask to closely follow a complex shape. Looking at the query in #3860, this is a similar example of what we are trying to do, it appears almost possible but the fact that there is no Z-plane means we apparently can only mask a slither and not a fully extruded and tight mask to follow a classifier.

It is possible for a CAD model, how feasible is it to make it so with a Reality Mesh model? ie add a 3rd dimension to a mask.

![ex](https://user-images.githubusercontent.com/108276006/175940876-53cd74ee-cc7c-4204-9d1a-e356e6715cc5.jpg)

This functionality is important to our workflow so i hope there is a solution.

Thanks. Tony

**Comment by pmconne** [comment]
There is no such thing as a "3d mask" - a mask is inherently 2d. As described above and more recently in #3860 you would need to define a 3d clip volume instead.

**Comment by tonygem2022** [comment]
Thank-you for the timely response.

The usefulness of a 3D clip in a congested environment, when trying to view selected components is limited. 

Would this be a standard response in a CAD environment? I think not. The example that @Rycha2021 is still open.

Isolating / segmenting specific elements in a reality model is equally important during our Client discussions.

The attempted solution per @bbastings early in the thread is close but when you have a multi-tiered pipe rack and the user needs to focus on a system meandering in multiple directions in different planes, the cube becomes ineffective. In a CAD environment this is possible. 

Therefore, will it be possible to consider a tight mask? If the pipe was duplicated with a large diameter to match the centreline of the pipe, why could there not be a viable solution to clip the shape accordingly?

Thanks. Tony

**Comment by pmconne** [comment]
> will it be possible to consider a tight mask

Masks are 2d. You are asking for a 3d solution. Masks are not your solution.

> The usefulness of a 3D clip in a congested environment, when trying to view selected components is limited

Please elaborate upon the perceived limitations.

> the cube becomes ineffective

Then use geometry more complex than a cube.

**Comment by tonygem2022** [comment]
Well Paul, if we had the below in a reality mesh model;

![image](https://user-images.githubusercontent.com/108276006/175947912-45dd8342-760f-4c61-9c38-c6346f9ac4a1.png)

and we wanted to see only this single line;

![image](https://user-images.githubusercontent.com/108276006/175948169-c7772c4c-70bd-4f35-ab0f-bbf803977464.png)

We couldnt. All we would see is this entire cube following your philosophy of clipping cubes as i understand.

![image](https://user-images.githubusercontent.com/108276006/175947912-45dd8342-760f-4c61-9c38-c6346f9ac4a1.png)

Now this is a rudimentary CAD model example, imagine in the example that #3860 has provided, having a cube just wont work as there will be cable trays, other piping, structure, small bore tubing / piping, other equipment in the field of view etc etc.

The outcome we really want to emulate is similar to a CAD environment and being able to select, track and only highlight the pipe from its inlet to its outlet in a Reality Mesh environment even if the route is multi-directional.

Perhaps 'Mask' is not the correct term but how can we emulate the 2D mask in 3 directions for complex shapes? I believe we perhaps can have similar functionality in Microstation but not iTwin? Effectively maybe the solution is a series of clipping cubes (or cylinders?). I don't know, but only brainstorming with you.

Thanks. Tony

**Reply by pmconne** [reply]
Thanks. A [ClipVector](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/clipvector/) is not limited to cubes - it can be arbitrarily complex. You could, for example, obtain the geometry of your element(s) as polyfaces and produce a ClipVector from them (one clip plane per facet). You would want to limit the level of detail to reduce the number of planes, and probably expand it slightly beyond the elements' volumes.

Is that clear and does that describe a solution to your problem?

**Reply by ovaisa** [reply]
In addition to the requirement outlined by @tonygem2022 (also in #3860) we have another important requirement which we were planning to use '3D mask'.  The requirement is to 'remove' an equipment mesh in the reality model to visualize the space without the equipment and for construction planning.  Hopefully these screen shots below will help understand our requirement

Original model

![Before-mask](https://user-images.githubusercontent.com/83007711/175955192-5cca3e85-b52e-479f-b1ab-f1a168d1ccba.JPG)

After 'removing' the equipment

![After-mask](https://user-images.githubusercontent.com/83007711/175955278-d442b8d6-c20b-4a56-a3cc-2e9eca0ff365.JPG)

We were planning to use mask capability to 'remove' the mesh, but we now realize that mask in its current form will not be able to achieve this objective.  Can you please advise how we might be able to achieve this requirement.

Thank you,

Ovais

**Reply by pmconne** [reply]
@ovaisa create ClipVector from mesh and invert it.

It seems all participants in this and related discussion want to be able to clip 3d volumes of a reality (or other) model based on the geometry of one or more elements. Agree?
If so, the solution will never be provided by masks --it will require a solution based on ClipVectors.

**Comment by tonygem2022** [comment]
"Thanks. A [ClipVector](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/clipvector/) is not limited to cubes - it can be arbitrarily complex. You could, for example, obtain the geometry of your element(s) as polyfaces and produce a ClipVector from them (one clip plane per facet). You would want to limit the level of detail to reduce the number of planes, and probably expand it slightly beyond the elements' volumes.

@pmconne. Is that clear and does that describe a solution to your problem?"

Unsure. Example would help. I will need to discuss with tech / dev team.

The solution needs to be enterprise, i.e. not hard coded for specific elements. 

I would hope the solution would align with the ease of use of the Classifier approach and toggle on/off in a similar way when the User selects (single or multiple elements i.e. user selects pipe, elbow, pipe) and then can clip the mesh accordingly.

Unsure if iTwin will pursue this capability but would support users in complex environments.

Thanks Tony

**Comment by Rycha2021** [comment]
Coming in late. Interesting that another similar request has been made. Thanks for pushing the same concept @tonygem2022.

We tried to do what was originally suggested but the results were not what we wanted readily and moved to other priorities. This Clip Vector approach was not an earlier suggestion so perhaps the thinking has evolved.....

Unclear why iTwin doesnt appear to consider this a 'core' functionality. Having the ability to segment 3D models, particularly mesh models, is a pivotal need to contextualise complex areas. I would hope this would be pushed higher up the dev road map.

@pmconne, your earlier comment "_I'm looking into a way to produce the effect you want. I don't think it will be difficult._" is probably more difficult than was originally conceived.

@tonygem2022, if you do manage to resolve this, kindly consider sharing.

@pmconne unsure how we add this t the Dev request list?

Ry

**Reply by pmconne** [reply]
My earlier comment "I don't think it will be difficult" was about masks and was implemented.
@markschlosseratbentley to add to backlog and evaluate scheduling.

**Reply by markschlosseratbentley** [reply]
Backlog item has been added, categorized as proposed: https://github.com/iTwin/itwinjs-core/issues/3883
Feel free to edit with more details.

**Comment by bbastings** [comment]
I did a little experiment to create a ClipVector from a ClipPrimitive constructed from a [UnionOfConvexClipPlaneSets](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/unionofconvexclipplanesets/).

`const clip = ClipVector.createCapture([ClipPrimitive.createCapture(planeSets)]);`

Model with no clipping:
![image](https://user-images.githubusercontent.com/65233531/176451957-74c787ae-e70c-4c25-b563-276210935b0a.png)

Model after clipping:
![image](https://user-images.githubusercontent.com/65233531/176452096-6a198ff4-c0f1-4f6d-9c86-6956ddd5d133.png)

Model clipping showing each ConvexClipPlaneSet:
![image](https://user-images.githubusercontent.com/65233531/176452449-2fc77fb3-f098-4857-a82f-516fe98d0123.png)

So, if you have something like a center line and radius it should be pretty easy for you to create a ConvexClipPlaneSet for each "segment" by creating ClipShape and asking the ClipShape for it's clip planes by calling "fetchClipPlanesRef". You then just need to add these to the UnionOfConvexClipPlaneSets using "addConvexSet".

HTH

**Comment by Rycha2021** [comment]
Interesting @bbastings. It appears to be along the lines of what we have asked.
Are you able to share a sandbox and/or code please? We could try and emulate it in our reality environment.
Thanks in advance. Ry

**Reply by ovaisa** [reply]
Request the same...

Thank you!

**Comment by tonygem2022** [comment]
@Rycha2021 no we havent progressed this. Whilst we want this towards the top of our list, our skill level is not where we could resolve this without copious amounts of trial and error dev.

Like ovaisa, we hopefully will receive the code that @bbastings could possibly share.

Thanks Tony.

**Comment by bbastings** [comment]
Here's a function to create a ClipVector from an array of points defining a path, a radius, and number of side.

`function createPathClip(points: Point3d[], radius: number, nSides: number): ClipVector | undefined {
  if (points.length < 2)
    return undefined;

  // Get polygon shape points (xy only) for given radius and number of sides...
  const xyPts: Point3d[] = [];
  const delta = (Math.PI * 2.0) / nSides;
  let angle = 0;
  const center = Point3d.create();
  const tmp = Point3d.create();
  for (let iPoly = 0; iPoly < nSides; iPoly++, angle += delta) {
    tmp.x = radius * Math.cos(angle);
    tmp.y = radius * Math.sin(angle);
    tmp.z = 0;
    xyPts.push(center.plus(tmp));
  }
  if (xyPts.length < 3)
    return undefined;

  // Create empty UnionOfConvexClipPlaneSets...
  const planeSets = UnionOfConvexClipPlaneSets.createEmpty();

  // Create a ClipShape for each segment of path as easy way to get a ConvexClipPlaneSet...
  for (let iPt = 0; iPt < points.length - 1; ++iPt) {
    const dir = Vector3d.createStartEnd(points[iPt], points[iPt + 1]);
    const matrix = Matrix3d.createRigidHeadsUp(dir);
    const transform = Transform.createOriginAndMatrix(points[iPt], matrix);

    const zLow = transform.multiplyInversePoint3d(points[iPt]);
    const zHigh = transform.multiplyInversePoint3d(points[iPt + 1]);

    const segmentClip = ClipShape.createShape(xyPts, zLow?.z, zHigh?.z, transform);
    if (undefined === segmentClip)
      return undefined;

    const segmentPlanes = segmentClip.fetchClipPlanesRef();
    if (undefined === segmentPlanes)
      return undefined;

    for (const convexSet of segmentPlanes.convexSets)
      planeSets.addConvexSet(convexSet);
  }

  // Create ClipVector from ClipPrimitive created from UnionOfConvexClipPlaneSets...
  const primitive = ClipPrimitive.createCapture(planeSets);
  return ClipVector.createCapture([primitive]);
}`

^ Add code doesn't seem to work correctly in this editor?

You can apply this to a view using: ViewClipTool.setViewClip(this.targetView, createPathClip(this.points, 1.0, 8));

HTH

-B
